experimental/segmenter/src/lstm_bies.rs

```diff
@@ -125,30 +147,63 @@ impl<'l> Lstm<'l> {
         &self,
-        x_t: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>,
-        h_tm1: &Array1<f32>,
-        c_tm1: &Array1<f32>,
-        warr: ArrayBase<ViewRepr<&f32>, Dim<[usize; 2]>>,
-        uarr: ArrayBase<ViewRepr<&f32>, Dim<[usize; 2]>>,
-        barr: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>,
+        x_t: MatrixBorrowed<'a, 1>,
+        mut h_tm1: MatrixBorrowedMut<'a, 1>,
+        mut c_tm1: MatrixBorrowedMut<'a, 1>,
+        warr: MatrixBorrowed<'a, 3>,
+        uarr: MatrixBorrowed<'a, 3>,
+        barr: MatrixBorrowed<'a, 2>,
         hunits: usize,
-    ) -> (Array1<f32>, Array1<f32>) {
-        // i, f, and o respectively stand for input, forget, and output gates
-        let s_t = x_t.dot(&warr) + h_tm1.dot(&uarr) + barr;
-        let i = math_helper::sigmoid_arr1(s_t.slice(ndarray::s![..hunits]));
-        let f = math_helper::sigmoid_arr1(s_t.slice(ndarray::s![hunits..2 * hunits]));
-        let _c = math_helper::tanh_arr1(s_t.slice(ndarray::s![2 * hunits..3 * hunits]));
-        let o = math_helper::sigmoid_arr1(s_t.slice(ndarray::s![3 * hunits..]));
-        let c_t = i * _c + f * c_tm1;
-        let h_t = o * math_helper::tanh_arr1(c_t.view());
-        (h_t, c_t)
+    ) -> Option<()> {
+        #[cfg(debug_assertions)]
+        {
+            let embedd_dim = x_t.dim();
+            h_tm1.as_borrowed().debug_assert_dims([hunits]);
+            c_tm1.as_borrowed().debug_assert_dims([hunits]);
+            warr.debug_assert_dims([hunits, 4, embedd_dim]);
+            uarr.debug_assert_dims([hunits, 4, hunits]);
+            barr.debug_assert_dims([hunits, 4]);
+        }
+
+        let mut s_t = barr.to_owned();
+
+        s_t.as_mut().add_dot_3d(x_t, warr);
+        s_t.as_mut().add_dot_3d(h_tm1.as_borrowed(), uarr);
+
+        for i in 0..hunits {
+            let tuple = s_t
+                .as_borrowed()
+                .submatrix::<1>(i)
+                .and_then(|s| s.read_4())
+                .unwrap_or((0.0, 0.0, 0.0, 0.0));
+            let p = math_helper::sigmoid(tuple.0);
+            let f = math_helper::sigmoid(tuple.1);
+            let c = math_helper::tanh(tuple.2);
+            let o = math_helper::sigmoid(tuple.3);
+            let c_old = c_tm1.as_borrowed().as_slice().get(i)?;
+            let c_new = p * c + f * c_old;
+            *c_tm1.as_mut_slice().get_mut(i)? = c_new;
+            *h_tm1.as_mut_slice().get_mut(i)? = o * math_helper::tanh(c_new);
+        }
+        Some(())
     }
 
     /// `word_segmenter` is a function that gets a "clean" unsegmented string as its input and returns a BIES (B: Beginning, I: Inside, E: End,
     /// S: Single) sequence for grapheme clusters. The boundaries of words can be found easily using this BIES sequence.
     pub fn word_segmenter(&self, input: &str) -> String {
+        match self.word_segmenter_inner(input) {
+            Some(s) => s,
+            None => {
+                debug_assert!(false);
+                String::new()
```

Comment:
Is the empty `String` a valid return value? I'd expect it to have the same length as the input

Reply:
Yeah I guess a better return value would be a string with all S which is a valid BIES string